client/drain.go

```diff
@@ -0,0 +1,80 @@
+package client
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+func (c *Client) DrainSelf() error {
+	drainSpec := c.GetConfig().Drain
+	if drainSpec == nil {
+		return nil
+	}
+
+	logger := c.logger.With("drain")
+
+	now := time.Now()
+	drainReq := &structs.NodeUpdateDrainRequest{
+		NodeID: c.NodeID(),
+		DrainStrategy: &structs.DrainStrategy{
+			DrainSpec: *drainSpec,
+			StartedAt: now,
+		},
+		MarkEligible: false,
+		Meta:         map[string]string{"message": "shutting down"},
+		WriteRequest: structs.WriteRequest{
+			Region: c.Region(), AuthToken: c.secretNodeID()},
+	}
+	if drainSpec.Deadline > 0 {
+		drainReq.DrainStrategy.ForceDeadline = now.Add(drainSpec.Deadline)
+	}
+
+	var drainResp *structs.NodeDrainUpdateResponse
+	err := c.RPC("Node.UpdateDrain", drainReq, drainResp)
+	if err != nil {
+		logger.Error("failed to drain self on shutdown", "error", err)
+		return err
+	}
+
+	// note: letting this run forever seems wrong but init system (ex systemd)
+	// will almost always force kill the client eventually
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if drainSpec.Deadline > 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), drainSpec.Deadline)
+		defer cancel()
+	}
+
+	logger.Debug("monitoring self-drain")
+
+	statusCheckInterval := time.Second * 10
+	timer, stop := helper.NewSafeTimer(0)
+	defer stop()
+
+	statusReq := &structs.NodeSpecificRequest{
+		NodeID:   c.NodeID(),
+		SecretID: c.secretNodeID(),
+		QueryOptions: structs.QueryOptions{
+			Region: c.Region(), AuthToken: c.secretNodeID()},
+	}
+	var statusResp *structs.SingleNodeResponse
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Error("self-drain exceeded deadline")
+			return ctx.Err()
+		case <-timer.C:
+			err := c.RPC("Node.GetNode", statusReq, statusResp)
+			if err != nil {
+				logger.Error("could not check node status", "error", err)
```

Comment:
We really need our RPC layer to support cancellation, but I think it requires adopting a fork of `net/rpc` which way too much scope to add for this effort. :sob: 

Even without that I think we could do a couple improvements:

1. On `ctx.Done()` try the RPC once more to avoid spurious error logs when everything actually worked as desired (since the 10s wait leaves a pretty sizable window for the deadline to be hit before we do another check).
   - If we're worried about an additional RPC blocking for too long, what if we just checked to see if there are any non-terminal allocations locally? We'd have to manually implement the `ignore system job` logic here as well which is a bit gross, but I love the idea of letting Clients make _local_ decisions.
2. Disconnected nodes without a deadline will never shutdown cleanly. I think we need some escape hatch like `if 10 minutes have passed without a successful RPC, exit`. I would just hate large clusters with fairly automated operations to have orphaned nodes hanging around because they never expected them to block forever.

That being said I'm not sure 10 minutes is sufficiently low to be safe. We could just rely on the user configuring their init system to force kill Nomad after a timeout of their choosing. The docs for this configuration block would be a natural place to call that out.

Reply:
I really like the idea of having the logic all be local and it _mostly_ solves problem (2) as well. We don't retry the initial `Node.UpdateDrain`, so if we're already disconnected at the time we shutdown that's fine -- we just stop the allocrunners and leave the tasks up as usual (which allows our disconnected clients logic to work as well). So long as we don't disconnect in the narrow window between when we call `Node.UpdateDrain` and the first `Client.GetAllocs` call, we don't need to talk to the server any more after that. So instead of calling `Node.GetNode` we can poll the alloc runners with a check like:

```go
check := func() {
	c.allocLock.RLock()
	defer c.allocLock.RUnlock()
	for _, runner := range c.allocs {
		// ...
	}
}
```

The ignore-system logic is pretty trivial.

Reply:
@schmichael implementing local-only behavior as we discussed here was easy but then I realized it's easy because it doesn't work. :grinning: 

We only remove an allocrunner from `c.allocs` when the client has been notified by the server that we've GC'd on the server (ref [`client.go#L2540-L2541`](https://github.com/hashicorp/nomad/blob/v1.5.3/client/client.go#L2540-L2541)). We also can't simply track the local client status because ephemeral disk migration relies on receiving the client status update to the server before making the data migration HTTP call, so we're supposed to be waiting on that before considering the drain complete.

Given that all the above is waiting for external requests (either server-to-client RPC or client-to-client HTTP), maybe we poll the server for drain status as I did originally, and then if we _can't_ poll the server because we're disconnected, we treat that as being safe to fall back to checking local `ClientStatus` of the allocs.